### PR TITLE
OLS-581: Move the program logic to wait for OLS 

### PR DESCRIPTION
**Pull request by Nehanth**
## Type of change

- [ ] Refactor
- [ ] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library
- [ ] Bump-up library or tool used for development (does not change the final image)

## Related Tickets & Documents

- Related Issue #
- Closes #

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
- Please provide detailed steps to perform tests related to this code change.
- How were the fix/results from this change verified? Please provide relevant screenshots or results.

### CODE REVIEW

File: tests/e2e/test_api.py [hunk 1]
```diff
@@ -34,10 +35,14 @@
 from tests.e2e.utils.response_evaluation import ResponseEvaluation
 from tests.scripts.must_gather import must_gather
 
+sys.path.append(Path(__file__).parent.parent.parent.as_posix())
```
Comment (bparees) [inline tests/e2e/test_api.py:38]: not clear to me why this is needed?  we are already importing other `test.scripts` content without this

File: tests/e2e/test_api.py [hunk 2]
```diff
@@ -76,6 +81,14 @@ def setup_module(module):
 
         client = client_utils.get_http_client(ols_url, token)
         metrics_client = client_utils.get_http_client(ols_url, metrics_token)
+
+        # Wait for OLS to be ready
+        success = wait_for_ols(ols_url)
+        if not success:
+            generate_junit_report(os.getenv("SUITE_ID"), False)
```
Comment (bparees) [inline tests/e2e/test_api.py:88]: does the pytest.fail below not already guarantee a junit test entry being produced?  (put another way, what does the junit + test job output look like if you don't explicitly do this?)

File: tests/scripts/utils.sh [hunk 3]
```diff
@@ -100,13 +85,15 @@ function run_suite() {
...
   install_ols "$1" "$3" "$4" "$5" "$6" "$7" "$8" "$9"
 
-  # wait for the ols api server to come up
-  wait_for_ols "$OLS_URL"
-  if [ $? -ne 0 ]; then
-    echo "Timed out waiting for OLS to become available"
-    echo "${JUNIT_TEMPLATE}" | sed "s/SUITE_ID/${1}/g" > $ARTIFACT_DIR/junit_setup_${1}.xml
+ # Determine the hostname for the ols route
+  export OLS_URL=https://$(oc get route ols -o jsonpath='{.spec.host}')
+  echo "OLS_URL set to $OLS_URL"
```
Comment (bparees) [inline tests/scripts/utils.sh:90]: to be fair this PR does what the ticket asked, but would it be terribly difficult to also move this logic into the setup while you're doing the refactoring?

File: tests/scripts/wait_for_ols.py
Comment (bparees) [inline tests/scripts/wait_for_ols.py:1]: instead of treating this as a standalone "script", i would just move it into tests.e2e.utils (either in one of the existing files or a new one).  

Is there an expectation this will get invoked as a standalone script, as compared to just using the utility function(s) it provides?

File: tests/scripts/wait_for_ols.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        failure = ElementTree.SubElement(
+            testcase, "failure", message="OLS failed to start up"
+        )
+        failure.text = "OLS did not become available in time"
+    tree = ElementTree.ElementTree(testsuite)
+    artifact_dir = os.getenv("ARTIFACT_DIR", ".")
+    tree.write(f"{artifact_dir}/junit_setup_{suite_id}.xml")
+
+
+if __name__ == "__main__":
```
Comment (bparees) [inline tests/scripts/wait_for_ols.py:83]: similar to the comment above about moving this whole thing into/under our tests.e2e.utils module, does it actually need logic to be invokable as a standalone script?  It doesn't look like this PR is doing so.  I'm guessing you just created this while you were doing some debugging?

**Review by bparees — COMMENTED** [review]
a couple structural questions.

also multiple commits is fine while you're working on the PR but please be prepared to squash your commits prior to us merging this.  (We only retain separate commits in a PR if there is a good reason to keep certain changes in their own commits, but if it takes you 10 commits to iterate on a solution, we don't keep those separate when we merge)

File: tests/e2e/test_api.py (hunk 1 above)
Comment (Nehanth) [reply tests/e2e/test_api.py:38]: it fails in CI If I do not do that for some reason

Comment (bparees) [reply tests/e2e/test_api.py:38]: @tisnik maybe you can help understand/explain why this is needed?

File: tests/e2e/test_api.py (hunk 2 above)
Comment (Nehanth) [reply tests/e2e/test_api.py:88]: oh okay, Will just remove it then

File: tests/e2e/utils/wait_for_ols.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import sys
+import time
+import warnings
+from pathlib import Path
+
+import requests
+from requests.exceptions import SSLError
+from urllib3.exceptions import InsecureRequestWarning
+
+sys.path.append(Path(__file__).resolve().parents[1].as_posix())
```
Comment (bparees) [inline tests/e2e/utils/wait_for_ols.py:15]: another thing i would not expect to be needed, especially now that this is a utility/helper, not a standalone script invocation.

File: tests/e2e/test_api.py [hunk 6]
```diff
@@ -74,8 +77,24 @@ def setup_module(module):
         else:
             print("Setting up for standalone test execution\n")
 
+        # Determine the hostname for the OLS route
+        ols_url = os.popen("oc get route ols -o jsonpath='{.spec.host}'").read().strip()
```
Comment (bparees) [inline tests/e2e/test_api.py:81]: we've got some helper functions for interacting with `oc` under "cluster.py", please take a look at that so we can do this in a consistent way

File: tests/scripts/utils.sh [hunk 7]
```diff
@@ -100,15 +85,10 @@ function run_suite() {
   
   install_ols "$1" "$3" "$4" "$5" "$6" "$7" "$8" "$9"
 
-  # wait for the ols api server to come up
-  wait_for_ols "$OLS_URL"
-  if [ $? -ne 0 ]; then
-    echo "Timed out waiting for OLS to become available"
-    echo "${JUNIT_TEMPLATE}" | sed "s/SUITE_ID/${1}/g" > $ARTIFACT_DIR/junit_setup_${1}.xml
```
Comment (bparees) [inline tests/scripts/utils.sh:107]: JUNIT_TEMPLATE(actually apparently it was JUNIT_SETUP_TEMPLATE) isn't used anymore, so you can delete it.

But fundamentally also the junit failure that JUNIT_TEMPLATE was used to generate, needs to be replicated by your new logic (i.e. we should still see a junit test failure representing the fact that OLS didn't become ready)

ironically also i think there was a bug here because it should have been JUNIT_SETUP_TEMPLATE, not JUNIT_TEMPLATE.  In any case the template is not needed now because replacing that template/generated junit test result is part of what your work is intended to replace.

File: tests/e2e/test_api.py [hunk 8]
```diff
@@ -55,14 +57,24 @@
...
 OLS_COLLECTOR_DISABLING_FILE = OLS_USER_DATA_PATH + "/disable_collector"
 
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+logger = logging.getLogger(__name__)
+
 
+# ruff: noqa: S605, S607
```
Comment (tisnik) [inline tests/e2e/test_api.py:60]: it is usually better to put these comments after lines with "problematic" commands

File: tests/e2e/test_api.py (hunk 6 above)
Comment (tisnik) [reply tests/e2e/test_api.py:81]: it might be saver to use `subprocess.Popen` with all the error checking etc.

File: tests/e2e/test_api.py [hunk 9]
```diff
@@ -72,13 +84,41 @@ def setup_module(module):
...
+            logger.info("Setting up for standalone test execution")
             print("Setting up for standalone test execution\n")
 
+        # Determine the hostname for the OLS route
+        ols_url = os.popen("oc get route ols -o jsonpath='{.spec.host}'").read().strip()
+        if not ols_url.startswith("http"):
+            ols_url = f"https://{ols_url}"
+
+        logger.info(f"OLS_URL set to {ols_url}")
+        print(f"OLS_URL set to {ols_url}")
```
Comment (tisnik) [inline tests/e2e/test_api.py:85]: log/print the same thingy - needed?

File: tests/e2e/test_api.py [hunk 10]
```diff
@@ -72,13 +84,41 @@ def setup_module(module):
...
+        os.environ["OLS_URL"] = ols_url
+
         client = client_utils.get_http_client(ols_url, token)
         metrics_client = client_utils.get_http_client(ols_url, metrics_token)
+
+        # Wait for OLS to be ready
+        logger.info(f"Waiting for OLS to be ready on provider: {provider}...")
+        print(f"Waiting for OLS to be ready on provider: {provider}...")
```
Comment (tisnik) [inline tests/e2e/test_api.py:92]: dtto

File: tests/e2e/test_api.py [hunk 11]
```diff
@@ -72,13 +84,41 @@ def setup_module(module):
...
+        success = wait_for_ols(ols_url)
+        if not success:
+            failure_message = (
+                f"OLS did not become available in time on provider: {provider}"
+            )
+            logger.error(failure_message)
+            print(failure_message)
```
Comment (tisnik) [inline tests/e2e/test_api.py]: dtto

File: tests/e2e/test_api.py [hunk 12]
```diff
@@ -72,13 +84,41 @@ def setup_module(module):
...
+            must_gather()
+            pytest.exit(f"OLS did not become available in time on provider: {provider}")
+        logger.info("OLS is ready")
+        print("OLS is ready")
```
Comment (tisnik) [inline tests/e2e/test_api.py:98]: dtto

File: tests/e2e/test_api.py [hunk 13]
```diff
@@ -72,13 +84,41 @@ def setup_module(module):
...
+
     except Exception as e:
-        print(f"Failed to setup ols access: {e}")
-        sys.exit(1)
+        failure_message = f"Failed to setup ols access: {e}"
+        logger.error(failure_message)
+        print(failure_message)
```
Comment (tisnik) [inline tests/e2e/test_api.py]: dtto

File: tests/e2e/utils/wait_for_ols.py (hunk 5 above)
Comment (tisnik) [reply tests/e2e/utils/wait_for_ols.py:15]: hmm +1 tbh

File: tests/e2e/utils/wait_for_ols.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    attempts = int(timeout / interval)
+    for attempt in range(1, attempts + 1):
+        print(f"Checking OLS readiness, attempt {attempt} of {attempts}")
+        try:
+            response = requests.get(f"{url}/readiness", verify=True, timeout=5)
+            if response.status_code == requests.codes.ok:
+                print("OLS is ready")
+                return True
+        except SSLError:
+            print("SSL error detected, retrying without SSL verification")
```
Comment (tisnik) [inline tests/e2e/utils/wait_for_ols.py:42]: so does it mean that even for https it can skip SSL?

**Review by tisnik — COMMENTED** [review]
some comments

File: tests/e2e/conftest.py [hunk 15]
```diff
@@ -1,17 +1,55 @@
...
+    # If OLS did come up cleanly during setup, then just generate normal test reports for all tests
+    return TestReport.from_item_and_call(item, call)
+
 
 def pytest_addoption(parser):
     """Argument parser for pytest."""
     parser.addoption(
         "--eval_provider",
-        default="openai",
+        default="watsonx",
```
Comment (bparees) [inline tests/e2e/conftest.py:46]: not sure where this change came from but i don't think it should be part of this PR

File: tests/e2e/conftest.py [hunk 16]
```diff
@@ -1,17 +1,55 @@
...
         type=str,
         help="Provider name, currently used only to form output file name.",
     )
     parser.addoption(
         "--eval_model",
-        default="gpt",
+        default="ibm/granite-13b-chat-v2",
```
Comment (bparees) [inline tests/e2e/conftest.py:52]: not sure where this change came from but i don't think it should be part of this PR

File: tests/e2e/conftest.py [hunk 17]
```diff
@@ -26,12 +64,6 @@ def pytest_addoption(parser):
         default=None,
         help="Ids of questions to be validated. Check json file for valid ids.",
     )
-    parser.addoption(
-        "--eval_threshold",
-        type=float,
-        default=None,
-        help="Threshold value to be used for similarity score.",
-    )
```
Comment (bparees) [inline tests/e2e/conftest.py:34]: not sure where this change came from but i don't think it should be part of this PR

File: tests/e2e/pytest.ini [hunk 18]
```diff
@@ -8,4 +8,4 @@ markers =
 	any
 	cluster
 	rag
-	response_evaluation
+	response_evaluation
```
Comment (bparees) [inline tests/e2e/pytest.ini:11]: these look like inadvertent whitespace changes, should not be part of this PR

File: tests/e2e/test_api.py [hunk 19]
```diff
@@ -98,11 +123,12 @@ def check_content_type(response, content_type, message=""):
     assert response.headers["content-type"].startswith(content_type), message
 
 
+@retry(max_attempts=3, wait_between_runs=10)
 def test_readiness():
     """Test handler for /readiness REST API endpoint."""
     endpoint = "/readiness"
     with metrics_utils.RestAPICallCounterChecker(metrics_client, endpoint):
-        response = client.get(endpoint, timeout=BASIC_ENDPOINTS_TIMEOUT)
+        response = client.get(endpoint, timeout=LLM_REST_API_TIMEOUT)
```
Comment (bparees) [inline tests/e2e/test_api.py:131]: looks like you need to rebase your branch to get rid of some of these spurious changes.

File: tests/e2e/utils/wait_for_ols.py (hunk 5 above)
Comment (bparees) [reply tests/e2e/utils/wait_for_ols.py:15]: I still think this can be removed and should be

File: tests/e2e/utils/wait_for_ols.py (hunk 14 above)
Comment (bparees) [reply tests/e2e/utils/wait_for_ols.py:42]: it means it can skip verifying the certificate.  i don't think it's an issue for testing purposes, the OLS cert is signed by the cluster CA, so if we wanted to verify it we'd need to fetch the CA from the cluster and feed it into the test process.

File: tests/scripts/utils.sh [hunk 20]
```diff
@@ -100,15 +85,10 @@ function run_suite() {
...
-    echo "Timed out waiting for OLS to become available"
-    echo "${JUNIT_SETUP_TEMPLATE}" | sed "s/SUITE_ID/${1}/g" > $ARTIFACT_DIR/junit_setup_${1}.xml
-
-    ARTIFACT_DIR=$ARTIFACT_DIR SUITE_ID=$1 python tests/scripts/must_gather.py
-    return 1
-  fi
+  # Determine the hostname for the ols route
+  #export OLS_URL=https://$(oc get route ols -o jsonpath='{.spec.host}')
+  #echo "OLS_URL set to $OLS_URL"
+  export SUITE_ID=$1
```
Comment (bparees) [inline tests/scripts/utils.sh:91]: what's the purpose of this?  SUITE_ID is set and provided to the make invocation below.

**Review by bparees — COMMENTED** [review]
please rebase this on top of the latest `main` branch to get rid of some of the weird deltas that are showing up, and remove any other changes this PR introduces that are not specifically about migrating the wait_for_ols logic.

**Comment by Nehanth** [comment]
Close
